### ISSUE TITLE
Internal array not getting reset?

### ISSUE THREAD

**Issue by bmueller-sykes** [joshwcomeau/react-flip-move#7]
Hi,

Apologies if this is an ill-formed question/issue. I've been toying around with this library, which seems super cool and very smooth, and during one of my tests I've stumbled across something that is either a bad implementation on my part (for which I'm unclear how to resolve) or a bug in the library. I'll paste in my code below, but the gist is this: I created a simple grid, with a "shuffle" button which, well, shuffles the tiles. As a test, I wanted to change the `opacity` of each tile as it was moving, so I set a "moving" flag on the `onStart()` function, and then unset that flag on `onFinish()` (then, based on that flag, I set the `opacity` to 1 or 0.5. The first shuffle works very smoothly, but on subsequent shuffles, it took longer and longer for the `opacity` to reset. So I threw a `console.log` in the `render()` function, and noticed that with each click of the `shuffle` button, the number of renders kept going up, by the same number of tiles that moved each time. So, on a 3x3 grid, it initially rendered 9 times, then 16, then 23, and so on. In the code below, you'll notice that the number of tiles the code thinks it's rendering doesn't change, which is good--it's the total number of renders that keeps going up.

It kind of looks like the library is keeping track of the items that have moved (obviously, not every tile moves on each shuffle, hence the reason it's not 9, 18, 27, etc), and isn't resetting that counter.

So I don't know if this is a feature or a bug. If it's the former, then I'm not sure how to reset it.

Also (bonus feature request), is it possible to have something like an `onAllFinish()` method? I could certainly see cases where I'd like to be able to fire off an event after all the animations are done.

Thanks!!

Code below:

    import React, { Component } from 'react';
    import FlipMove from 'react-flip-move';
    import _ from "lodash";

    export default class GridContainer extends Component {

	constructor (props) {
		super(props);
		
		
		this.config = {
			rows: 3,
			columns: 3,
			tileWidth: 100,
			tileHeight: 100,
			tileMarginRight: 2
		}
		
		this.config.numTiles = this.config.rows * this.config.columns;
		
		this.state = {
			tileData: _.times(this.config.numTiles, (index) => (
				{
					id: index,
					moving: false
				})
			)			
		}
	}

	handleShuffle = (evt) => {
		evt.preventDefault();
		
		this.setState({
			tileData: _.shuffle(this.state.tileData)
		})
	}

	handleMoveStart = (child, node) => {
		
		const tiles = this.state.tileData.slice(); //create new array to mutate
		const current = tiles.findIndex( (tile) => { return tile.id === child.ref; })

		tiles[current].moving = true;
		
		this.setState({
			tileData: tiles
		})	
	
	};
	
	
	handleMoveFinish = (child, node) => {
		const tiles = this.state.tileData.slice(); //create new array to mutate
		const current = tiles.findIndex( (tile) => { return tile.id === child.ref; })

		tiles[current].moving = false;
		
		this.setState({
			tileData: tiles
		})	
	};
	
	
    render() {
      return (
	  <div>
			<div><button onClick={this.handleShuffle}>Shuffle</button></div>
			<Grid config={this.config} tileData={this.state.tileData} handleMoveStart={this.handleMoveStart} handleMoveFinish={this.handleMoveFinish} />
	  </div>
      )
    }
    }

    class Grid extends Component {
	
	tiles = () => {
		
		return this.props.tileData.map ( (tile) => {
			return this.tile(tile);
		})
		
	};
	
	
	tile = (data) => {
		
		const style = {
			width: this.props.config.tileWidth,
			height: this.props.config.tileHeight,
			backgroundColor: "yellow",
			borderRadius: 5,
			float: "left",
			marginRight: this.props.config.tileMarginRight,
			marginBottom: 2,
			opacity: (data.moving) ? "0.5" : "1.0"
			//transition: "opacity 25 ease-in-out"
		}
		
		
		return (
			<div key={data.id} ref={data.id} style={style}>tile {data.id}</div>
		)		
		
	};

	
	
	render () {

		const gridWidth = (this.props.config.tileWidth * this.props.config.columns) + (this.props.config.tileMarginRight * this.props.config.columns);
		const gridHeight = this.props.config.tileHeight * this.props.config.rows;
		
		const gridStyle = {
			width: gridWidth,
			height: gridHeight
		}
		
		console.log("RENDERING NUM TILES:", this.props.tileData.length);
		
		return (
			<div style={gridStyle}>
				<FlipMove 
					duration={250}
					easing="ease-in-out"
					staggerDurationBy={10} 
					staggerDelayBy={0} 
					onStart={this.props.handleMoveStart}
					onFinish={this.props.handleMoveFinish}
				>
					{this.tiles()}
				</FlipMove>
			</div>
		)	
	}
    }

**Comment by joshwcomeau** [comment]
Interesting! I have a few ideas for what it could be.

Can you do me a favour and temporarily remove the calls to `this.setState()` in `handleMoveStart` and `handleMoveFinish`? And let me know if the number of `render()` calls is still increasing on every shuffle (and, if so, by the same amount?).

Either way, I'm pretty sure this is a bug in _my_ implementation, not yours :) I will try and get it sorted out in the next few days.

**Comment by bmueller-sykes** [comment]
For the record, it appeared in both 1.0.4 and 1.1.0. (I just realized I wasn't sure what version I was on, and wanted to make sure this wasn't just due to an old version).

yes, I'll try that test now.

**Comment by bmueller-sykes** [comment]
Okay, removing `setState()` in `handleMoveStart()` and `handleMoveFinish()` causes things to function normally--there's no explosion of render calls.

As I wrote this test code, I kind of thought to myself "this is probably not the right way to do this", but given that the goal is to abstract the tile data away from the UI data, putting a "moving" flag in the data external to the `<Grid>` call seemed like the right way to do things. And yet, I knew that each `setState()` call was going to force a new `Render` call. It does seem like 1 `Render` call per moving item could be stay performant, but obviously, an explosion of `Render` calls is not sustainable.

I wonder: is it at all possible to put some kind of an `isMoving` flag on each item at the library level, and then expose that? Certainly for CSS changes, just having a flag like that available as a prop to the child elements would be pretty useful. But I don't know if that's technically feasible or not.

**Comment by joshwcomeau** [comment]
Yeah, I suspected something like that was happening. 

Exposing some sort of `moving` flag in the library makes a ton of sense! I'm not sure what the best way to do that would be though. My first thought would be to add a class name directly, sorta like ReactCSSTransitionGroup:

```js
<FlipMove duration={300} movingClassName="moving">
  { this.stuff() }
</FlipMove>

// example rendered node: <div className="stuff active moving"></div>
```

I like this idea because I don't want to assume that all children passed to FlipMove are components; you might want to do it on plain DOM nodes, in which case the `isMoving` prop would be meaningless.

I think it's feasible from a technical standpoint but I won't know until I have the chance to dig in and try it.

Also, I didn't miss your suggestion for an `onAllFinish` callback. I was actually planning on making the `onFinish` callback act this way, until I realized that it was much easier to do it by-child rather than by-group. I figured I'd wait and see if there was demand before implementing it. Now that I know there is, I'll try and implement it shortly as well :)

**Comment by bmueller-sykes** [comment]
My understanding is that the React community in general is moving away from CSS in favor of inline style calls (for a couple of reasons, not the least of which is that inline style calls lend themselves more seamlessly to React Native), so if you want to allow for a "movingClassName", I would also encourage you to add an `isMoving` flag.

As for `onFinish` vs. `onAllFinish`, assuming that it can stay performant, I think there's a case to be made that developers would want to be able to call a function for each item *and* for everything, so I'd recommend having both.  Of course, if having `onFinish` will ultimately result in badly performant code, then maybe it makes sense to just not even allow for that.  But that's just my $.02.

**Comment by joshwcomeau** [comment]
Yeah, I meant I'd add an `onAllFinish` in addition to the `onFinish`. It would be horrible of me to change how an existing API method worked!!

And yeah, that's a good point about the CSS modules trend. I guess for most practical purposes, you'd be animating child components and not regular DOM elements. I'm still a bit resistant to mutating the children in such a way, but I'd only be doing it when it's requested, so I suppose it's not a big deal.

Gonna give it a bit more thought.

**Comment by joshwcomeau** [comment]
So, happily, someone else found the issue; I wasn't un-binding the `transitionEnd` event listener. This means that every time the animation finished, it was calling the supplied callback `n` extra times.

I accepted their PR and released a new version ; upgrade to v1.1.1 and your solution should work perfectly, and not degrade over time.

Now that this is fixed, I don't believe an `isMoving` flag is necessary, but I will definitely add the `onAllFinish` callback to the roadmap :)

**Comment by bmueller-sykes** [comment]
Confirmed! A few notes below, but before I forget, I wonder if in addition to `onAllFinish()`, I wonder if an `onAllStart()` might make sense as well.

I tested on a 2015 MacBookPro, and in Chrome, Safari and FF, the animation was super-smooth. Same with Safari, Chrome and FF on an iPhone 6s. IE11, Chrome and Firefox on Windows 8 running inside VMware on the MBP were a little jittery, though. I assume that's just because it's in a VM that isn't tuned for performance, but I suppose there's a small chance that the moves are jittery on a native Windows box. Might be worth testing, if you haven't done that already.

Thanks again!!

**Comment by joshwcomeau** [comment]
Awesome :)

I can think of a few use-cases for `onAllStart`, but they're pretty obscure (if you pass in a `delay`, or if you only want something to fire when the props are different enough to warrant an animation). I'm hesitant to add stuff to the API that doesn't have a common use-case, but if there's one I'm missing, please share!

I did test FlipMove in general on an older windows machine, in Chrome and IE10. The demo site looked all funky because flexbox wasn't supported, but the animation itself was smooth. That said, I never tested it with an onFinish callback, so that's probably worth a shot.

Thanks for testing it on different browsers though!

**Comment by joshwcomeau** [comment]
Hey @bmueller-sykes!

Just a quick note; I released Flip Move v1.3 which features an `onFinishAll` callback :)

Its signature is similar to `onFinish`, except it is provided arrays of elements/domNodes instead of individual ones.

Still no `onStartAll`; it doesn't strike me as 100% necessary, and I want to keep the API footprint as small as possible.

**Comment by bmueller-sykes** [comment]
#hero
